templates/default/postgres.yaml.erb

```diff
@@ -1,53 +1,21 @@
-init_config:
-# empty by design
+# Generated by Chef, local modifications will be overwritten
+
+<%# TODO: Breaking change, remove defaults -%>
+<%
+  # Apply default transformations to @instances, an ImmutableArray
+  instances = @instances.to_a
+
+  instances.each do |instance|
+    # The `server` key renders as a `host` value
+    # `if` test future-proofs the key from being modified if users change to
+    # using `host` as their key.
+    instance['host'] = instance.delete 'server' if instance['server']
 
-instances:
-<% @instances.each do |i| -%>
-  - host: <%= i["server"] %>
-    port: <%= i["port"] || 5432 %>
-    username: <%= i["username"] %>
-    password: <%= i["password"] %>
-    <% if i.key?("tags") -%>
-    tags:
-        <% i["tags"].each do |t| -%>
-      - <%= t %>
-      	<% end -%>
-    <% end -%>
-    <%- if i.key?('dbname') %>
-    dbname: <%= i['dbname'] %>
-    <% end -%>
-    <%- if i.key?('ssl') %>
-    ssl: <%= i['ssl'] %>
-    <% end -%>
-    <% if i.key?("relations") -%>
-    relations:
-        <% i["relations"].each do |r| -%>
-      - <%= r %>
-        <% end -%>
-    <% end -%>
-    <%- if i.key?("collect_function_metrics") %>
-    collect_function_metrics: <%= i["collect_function_metrics"] %>
-    <% end -%>
-    <% if i.key?('custom_metrics') -%>
-    custom_metrics:
-      <% i["custom_metrics"].each do |cm| -%>
-      - query: <%= cm["query"] %>
-        metrics:
-          <% cm["metrics"].each do |column, vals| -%>
-          <%= column %>:
-            <% Array(vals).each do |val| -%>
-            - <%= val %>
-            <% end -%>
-          <% end -%>
-        <% if cm.key?("descriptors") -%>
-        descriptors:
-          <% Array(cm["descriptors"]).each do |d| %>
-          - <%= d %>
-          <% end -%>
-        <% end -%>
-        <% if cm.key?("relation") -%>
-        relation: <%= cm["relation"] %>
-        <% end -%>
-      <% end -%>
-    <% end -%>
-<% end -%>
+    # If `port` is unspecified, set to 5432
+    instance['port'] = 5432 if instance['port'].nil?
+  end
+-%>
+<%= JSON.parse(({'instances' => instances}).to_json).to_yaml %>
```

Comment:
Nice hack for avoiding there being a second document (`---`) in this file. Hadn't even thought about putting that before the `init_config` line and was instead just building a Hash in the template file and rendering that to YAML.